sqsutil/prefetcher/prefetch.go

```diff
@@ -0,0 +1,94 @@
+package prefetcher
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+// Prefetcher is a struct that prefetches a value at a given interval
+// and provides a method to get the latest value.
+// NOTE: It may return stale data if the update function takes longer than the interval.
+type Prefetcher[T any] struct {
+	updateFn  func() T
+	interval  time.Duration
+	hasClosed bool
+
+	hasSuccesfullyFetched bool
+	firstFetchChan        chan struct{}
+	lastRetrievedTime     time.Time
+	cache                 T
+	timer                 *time.Ticker
+	mutex                 sync.RWMutex
+}
+
+func NewPrefetcher[T any](updateFn func() T, interval time.Duration) *Prefetcher[T] {
+	prefetcher := &Prefetcher[T]{
+		updateFn:       updateFn,
+		interval:       interval,
+		firstFetchChan: make(chan struct{}),
+	}
+
+	go prefetcher.startTimer()
+
+	return prefetcher
+}
+
+func (p *Prefetcher[T]) startTimer() {
+	p.prefetch()
+	p.timer = time.NewTicker(p.interval)
+
+	for range p.timer.C {
+		p.prefetch()
+	}
+}
+
+func (p *Prefetcher[T]) prefetch() {
+	newValue := p.updateFn()
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
```

Comment:
For data that has many key/value pairs and frequent access to various keys, would we have to maintain a separate copy under `sync.Map` somewhere so that we don't cause latency spikes when acquiring a mutex here?

Or maybe should we have a separate prefetcher implementation as a `sync.Map` for this access pattern?

Reply:
this should always be super quick, right?

This latency is the time it takes to either:
- copy a struct
- Swap a pointer (e.g. Slice or Map case)

So if you were querying a map, the overhead would be very small

Reply:
Yes, but situation changes under load. I've had production use cases where with moderate write load and heavy ready load this combination of map and RW lock would cause meaningful RPS degradation and tail latency spikes. `sync.Map` would perform much better in those cases.

I am, however, in agreement to not over-focus on this at this stage and observe the patterns once we start load testing and profiling

Reply:
If this was true, we just switch to a CAS instruction

Reply:
SGTM. I'm merging this